Add filename and line to functions when printed

ChrisJefferson opened the pull request:
This PR adds to functions their filename, and line, when they are printed (when the function comes from an actual file). As an example:

```
gap> Print(List);
# /home/caj/files/reps/gap/gap/lib/coll.gi:645
function ( arg... )
    local tnum, C, func, res, i, elm, l;
    ...
```

I put a `#` at the start so the printed out information is still a valid piece of GAP.

I'm interested if anyone thinks this is a good idea, or if there should be any other adjustments. I haven't gone through and done the (fairly mechanical) changes required to the tests yet.

fingolfin commented:
Actually, it is completly unclear to me how you plan to adjust the tests for this, given that paths will be different on every system. While I can think of a few approaches, none seems satisfying. So, what is the "mechanical" approach you have in mind?

ChrisJefferson commented:
Woops, you are right! My first thought was the boring bit of fixing the tests.

We've previously discussed some kind of way of marking the gap library root in filenames (something like `GAPROOT/lib/coll.gi`), which would make the path system-independant, at the cost of being harder to just cut+paste into an editor. If we didn't do that, then indeed I can't really see how to make this a valid patch.

fingolfin commented:
(Ignoring the technical concerns, I like the feature)

cdwensley commented:
It is a good feature - and is related, with the same problems, to issue #2871.

fingolfin commented:
One of the various way s one could address this in tests is to teach Test to transform outputs by replacing any occurrence of the GAP root path by the string GAPROOT, as we do in testspecial.

ChrisJefferson commented:
I could add a configuration option, which anyone can turn on/off, and we turn off when testing (I'd prefer to leave it on by default).

fingolfin commented:
But then we still need to ask all affected package authors to do that in their packages. But perhaps not that many are affected -- it'd be good to know how many before making any decision on that.

fingolfin commented on src/calls.c:
```diff
@@ -1105,6 +1105,16 @@ void PrintFunction (
+
+    if(filename && sl) {
+        // Filenames starting '*' are special streams
+        if(CHARS_STRING(filename)[0] != '*') {
```
```suggestion
        if (CHARS_STRING(filename)[0] != '*') {
```

stevelinton commented:
Maybe this should be part of a Display method for functions, rather than Print. That way it won't break existing tests, and it opens the way to adding more info if we think of any.

fingolfin commented:
In light of PR #2987 I think going with @stevelinton's suggestion might be the best course of action... ?

ChrisJefferson commented:
I've now moved the message to Display, moved the test to testspecial (where we already deal with filenames containing GAPROOT), and also stopped printing names that start with `*`, like `*stream*` and `*stdin*`. Obviously someone could have a filename that starts with a star, but it would be a silly name and there is a chance we already handle it badly somewhere.

alex-konovalov commented:
This may be useful. I am using `PageSource` for this, since it displays the location before displaying the source, but `Display` will be more convenient.
